pkg/volume/util/util.go

```diff
@@ -576,6 +576,38 @@ func IsLocalEphemeralVolume(volume v1.Volume) bool {
 		volume.ConfigMap != nil
 }
 
+// GetLocalPersistentVolumeNodeNames returns the node affinity node name(s) for
+// local PersistentVolumes. nil is returned if the PV does not have any
+// specific node affinity node selector terms and match expressions.
+// PersistentVolume with node affinity has select and match expressions
+// in the form of:
+//
+//	nodeAffinity:
+//	  required:
+//	    nodeSelectorTerms:
+//	    - matchExpressions:
+//	      - key: kubernetes.io/hostname
+//	        operator: In
+//	        values:
+//	        - <node1>
+//	        - <node2>
+func GetLocalPersistentVolumeNodeNames(pv *v1.PersistentVolume) []string {
+	if pv == nil || pv.Spec.NodeAffinity == nil || pv.Spec.NodeAffinity.Required == nil {
+		return nil
+	}
+	if len(pv.Spec.NodeAffinity.Required.NodeSelectorTerms) != 1 {
+		return nil
+	}
+
+	nodeSelector := pv.Spec.NodeAffinity.Required.NodeSelectorTerms[0]
```

Comment:
I rethink the logic a bit. Given NodeSelectorTerms are ORed:

```go
type NodeSelector struct {
    // Required. A list of node selector terms. The terms are ORed.
    NodeSelectorTerms []NodeSelectorTerm
}
```

and NodeSelectorTerm.MatchExpressions is ANDed. We should do something like:

```go
nodes := sets.String()
for each nodeSelectorTerm:
    for each matchExpression:
        if it matches {key: hostname, operator: In}:
            // calculate an intersected node set
    // at the end of each inner loop, union the calculated node set to the final `nodes`
```

Reply:
Thanks, done.